Fix reading of MBFITS tables with variable length array

cmarmo opened the pull request:
### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to address the issues of reading MBFITS tables with variable length arrays.
- The warning
```
WARNING: VerifyWarning: Invalid keyword for column 3: The repeat count of the column format 'USEFEED' for column '1PJ(8)' is fewer than the number of elements per the TDIM argument '(8,1)'.  The invalid TDIMn value will be ignored for the purpose of formatting this column. [astropy.io.fits.column]
WARNING: VerifyWarning: Invalid keyword for column 4: The repeat count of the column format 'BESECTS' for column '1PJ(8)' is fewer than the number of elements per the TDIM argument '(8,1)'.  The invalid TDIMn value will be ignored for the purpose of formatting this column. [astropy.io.fits.column]
WARNING: VerifyWarning: Invalid keyword for column 5: The repeat count of the column format 'FEEDTYPE' for column '1PJ(8)' is fewer than the number of elements per the TDIM argument '(8,1)'.  The invalid TDIMn value will be ignored for the purpose of formatting this column. [astropy.io.fits.column]
```
is silenced, taking into account the `recformat` attribute `max` rather than `repeat` for `Q` and `P` formats.

- Some constraints on dimensions obtained from `TDIM` have been relaxed  for `Q` and `P` formats.

- The test use the file linked in #7810

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

The PR fixes #7810.
It should also fix #12860.

Thanks for considering it.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

pllim commented:
Thanks! Would need a change log too. Please see https://github.com/astropy/astropy/blob/main/docs/changes/README.rst

saimn commented on astropy/io/fits/column.py:
```diff
@@ -1119,7 +1119,11 @@ def _verify_keywords(cls, name=None, format=None, unit=None, null=None,
+                    repeat = int(recformat.max)
+                else:
+                    repeat = format.repeat
+                if reduce(operator.mul, dims_tuple) > repeat:
```
So this is to fix the warning with the second file from #7810 right ? (`S60mm-ICPBE-FEBEPAR.fits`)

saimn replied:
It seems TDIM is not taken into account though ?
```
In [7]: hdulist = fits.open('S60mm-ICPBE-FEBEPAR.fits')
   ...: data = hdulist[1].data

In [8]: data['USEFEED']
Out[8]: _VLF([array([1, 1, 1, 1, 2, 2, 2, 2], dtype=int32)], dtype=object)

In [9]: data.columns['USEFEED']
Out[9]: name = 'USEFEED'; format = '1PJ(8)'; dim = '(8,1)'
```
(this is with your branch)

cmarmo replied:
> So this is to fix the warning with the second file from #7810 right ? (`S60mm-ICPBE-FEBEPAR.fits`)

Correct.

cmarmo replied:
> It seems TDIM is not taken into account though ?
> ...
> (this is with your branch)

I'm having different results.... but just to be sure I have added a test about TDIM.

saimn reviewed (commented):
Sorry @cmarmo it took me time to go through this time, remembering the issue etc. But I think there is still an issue with TDIM. Columns like USEFEED should have more dimensions, as is the case for e.g. APEREFF which is not a VLA and also has TDIM=(1, 1).

```
In [18]: data['USEFEED']
Out[18]: _VLF([array([1], dtype=int32)], dtype=object)

In [19]: data['APEREFF']
Out[19]: array([[[0.53]]], dtype=float32)
```
In https://github.com/astropy/astropy/issues/7810#issuecomment-421499877 you can see the result of a change I tried in  #7820 and where USEFEED has the missing dimensions. Maybe this change is actually needed ?

saimn commented:
@bwinkel - Is it OK to use the files you added in #7810 to check that the bug is fixed ? Did you generate those files, or if they come from somewhere else is it OK to distribute them in Astropy which is BSD licensed ?

bwinkel commented:
> @bwinkel - Is it OK to use the files you added in #7810 to check that the bug is fixed ? Did you generate those files, or if they come from somewhere else is it OK to distribute them in Astropy which is BSD licensed ?

The files comes from the Effelsberg 100-m radio telescope. I think, there should be no problem at all to use it for this purpose, but let me double check with my boss.

cmarmo commented:
I have rebased on main because of conflicts and added a test about `TDIM`: is `'(1,1)'` the expected result? Thanks!

bwinkel commented:
> > @bwinkel - Is it OK to use the files you added in #7810 to check that the bug is fixed ? Did you generate those files, or if they come from somewhere else is it OK to distribute them in Astropy which is BSD licensed ?
> 
> The files comes from the Effelsberg 100-m radio telescope. I think, there should be no problem at all to use it for this purpose, but let me double check with my boss.

Dear all, you can happily use the file. Thanks for fixing this bug!

saimn commented:
> I have rebased on main because of conflicts and added a test about TDIM: is '(1,1)' the expected result? Thanks!

TDIM is set to `(1, 1)` in the header, so that means that each element of e.g. USEFEED should be a 1x1 array, same as APEREFF for example. But now we have this:
```
In [18]: data['USEFEED']
Out[18]: _VLF([array([1], dtype=int32)], dtype=object)

In [19]: data['APEREFF']
Out[19]: array([[[0.53]]], dtype=float32)
```
where TDIM has not been applied to USEFEED. We should have something like `_VLF([array([[[1]]], dtype=int32)], dtype=object)` instead (which I was able to get with #7820, see the code example in https://github.com/astropy/astropy/issues/7810#issuecomment-421499877, though this change was not enough and maybe combined with your changes here it would work ?).

cmarmo commented:
Eventually, I understood your comment ... 😖
That means my test is bad too ....
Working on it and adding your change.

cmarmo commented:
Digging deeper I realized that my solution was wrong: I didn't trust the standards... :)
I have reconsidered and I'm more confident now: that's why I have rebased and squashed some commits.... the history was really messy.
I have incorporated your change, @saimn, as it is necessary to not reshape the values of the array descriptor.

cmarmo commented on astropy/io/fits/column.py:
```diff
@@ -1119,7 +1119,9 @@ def _verify_keywords(cls, name=None, format=None, unit=None, null=None,
 
             if dims_tuple:
-                if reduce(operator.mul, dims_tuple) > format.repeat:
+                if isinstance(recformat, _FormatP):
+                    msg = None
```
This is to avoid the warning for P and Q format as TDIM has a different meaning in VLA formats

cmarmo commented on astropy/io/fits/column.py:
```diff
@@ -1603,7 +1604,7 @@ def dtype(self):
                 offsets.append(offsets[-1] + dt.itemsize)
 
-            if dim:
+            if dim and format_.format not in 'PQ':
```
This avoid reshaping the array descriptor, which is always of shape (2,)

cmarmo commented on astropy/io/fits/fitsrec.py:
```diff
@@ -814,6 +816,8 @@ def _convert_p(self, column, field, recformat):
                 dt = np.dtype(recformat.dtype)
                 arr_len = count * dt.itemsize
                 dummy[idx] = raw_data[offset:offset + arr_len].view(dt)
+                if column.dim:
+                    dummy[idx] = dummy[idx].reshape(vla_shape)
```
Here the VLA is reshaped consistently with TDIM instructions.

cmarmo replied:
I realized that this is working in this specific case but, as VLA have variable length, this may not work when different rows have different length.... I'm working on figure this out, and adding a related test too.

cmarmo commented:
Working around https://github.com/astropy/astropy/pull/13417#discussion_r1015789251, I hit similar issues as in #12860.
I believe the two issues should indeed be solved together.
I am currently stuck trying to understand why I can't write the value I want into the VLF descriptor.
Running this code
```py
import numpy as np
from astropy.io import fits
a = np.ones((5, 2))
b = np.ones((7, 2))
array = np.array([a, b], dtype=object)
col = fits.Column(name='test', format='PD(14)', dim='(2,7)', array=array)
print("Write file")
fits.BinTableHDU.from_columns([col]).writeto('test.fits', overwrite=True)

print("\nRead file")
with fits.open('test.fits') as hdus:
    print("Column format ", hdus[1].columns.formats)
    print(hdus[1].data['test'])
```
with ca8ee5071930fae23433455f4155f95a0ba91999 gives
```pytr
Write file
write descriptor  [10  0]

Read file
Column format  ['PD(14)']
read descriptor  [5 0]
Traceback (most recent call last):
  File "/home/cmarmo/software/tests/12860-astropy.py", line 13, in <module>
    print(hdus[1].data['test'])
  File "/home/cmarmo/software/astropy/astropy/io/fits/fitsrec.py", line 501, in __getitem__
    return self.field(key)
  File "/home/cmarmo/software/astropy/astropy/io/fits/fitsrec.py", line 716, in field
    converted = self._convert_p(column, field, recformat)
  File "/home/cmarmo/software/astropy/astropy/io/fits/fitsrec.py", line 824, in _convert_p
    dummy[idx] = dummy[idx].reshape(vla_shape)
ValueError: cannot reshape array of size 5 into shape (2,7)
```
The failure is expected (the reshaping is not fixed yet): but I can't understand why I am supposed to write `[10 0]` in the file and when reading it I retrieve `[5 0]` ... 
@saimn , if you have any hint, when you are available, they will be really appreciated... Thank you very much.

cmarmo commented on astropy/io/fits/fitsrec.py:
```diff
@@ -1140,7 +1147,7 @@ def _scale_back(self, update_heap_pointers=True):
                     # update the array descriptors
                     raw_field[:] = 0  # reset
-                    npts = [len(arr) for arr in self._converted[name]]
+                    npts = [np.prod(arr.shape) for arr in self._converted[name]]
```
Here is where the descriptor is recalculated just before to be written to the file.

cmarmo commented:
I believe this one is finally ready for review.
Please note that #7810 is a consequence of #12860.
Now that multidimensional variable length array table can be correctly written, there is no longer need for the test using the MBFITS file.
Thanks for your patience.

cmarmo commented on astropy/io/fits/fitsrec.py:
```diff
@@ -837,6 +840,10 @@ def _convert_p(self, column, field, recformat):
                 arr_len = count * dt.itemsize
                 dummy[idx] = raw_data[offset : offset + arr_len].view(dt)
+                if column.dim and len(vla_shape) > 1:
+                    vla_dimx = vla_shape[0]
+                    vla_dimy = int(len(dummy[idx]) / vla_shape[0])
+                    dummy[idx] = dummy[idx].reshape(vla_dimx, vla_dimy)
```
this still need to be generalized to more than two dimensions.

cmarmo commented:
PR history cleaned, synchronized with main, tests updated.
Ready for review :rocket:

saimn commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -3270,6 +3270,72 @@ def test_empty_vla_raw_data(self):
+        and https://github.com/astropy/astropy/issues/7810
+        """
+        a = np.ones((5, 1))
+        b = np.ones((7, 1))
```
Can you test with values that are not all 1, to be sure that they are correctly reshaped, saved etc. ?

cmarmo replied:
Done.

saimn reviewed (commented):
Wow great work @cmarmo !
For now I just have a small request for the tests values. I will try to do other tests locally soon.

cmarmo commented:
> Wow great work @cmarmo !

Thanks! ☺️

> For now I just have a small request for the tests values. I will try to do other tests locally soon.

I will apply your comments ASAP... I'm not well connected these days...

saimn reviewed (approved):
Thanks @cmarmo :tada:

saimn commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -3270,6 +3270,72 @@ def test_empty_vla_raw_data(self):
+                        ]
+                    ),
+                ],
+            )
```
@cmarmo - it seems there is a problem here. This test is failing on the backport on v5.0.x (#14183, https://github.com/astropy/astropy/actions/runs/3715605304/jobs/6300973451), with an old version of Numpy. I investigated a bit, and found several problems:
- the test here should fail, it seems the data should be transposed. If I read `"test2.fits"` I get this:
```
In [3]: hdus = fits.open('test2.fits')

In [4]: hdus[1].data['test']
Out[4]: 
_VLF([array([[0., 1., 2., 3., 4.],
             [5., 6., 7., 8., 9.]]),
      array([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.],
             [ 7.,  8.,  9., 10., 11., 12., 13.]])], dtype=object)
```
which is not the expected result. 
- I spend some time trying to find why the test was not failing on `main` until I realized, you are not testing the result of `np.array_equal` :grimacing: . The test code should use `assert np.array_equal` (or `np.testing.assert_array_equal`).
- the test case above (`test_empty_vla_raw_data`) is also missing `assert` in front of `np.array_equal`, and is also failing if I add the `assert`. This one is from  #13621.

I should have seen that so no blame on you, but it seems we now have a few issues to deal with.

saimn replied:
Also, the way the test is written raises a warning from Numpy: 
```
In [5]:             np.array_equal(
   ...:                 hdus[1].data["test"],
   ...:                 [
   ...:                     np.array(
   ...:                         [[0.0, 1.0], [2.0, 3.0], [4.0, 5.0], [6.0, 7.0], [8.0, 9.0]]
   ...:                     ),
   ...:                     np.array(
   ...:                         [
   ...:                             [0.0, 1.0],
   ...:                             [2.0, 3.0],
   ...:                             [4.0, 5.0],
   ...:                             [6.0, 7.0],
   ...:                             [8.0, 9.0],
   ...:                             [10.0, 11.0],
   ...:                             [12.0, 13.0],
   ...:                         ]
   ...:                     ),
   ...:                 ],
   ...:             )
/home/simon/.pyenv/versions/3.9.8/lib/python3.9/site-packages/numpy/core/numeric.py:2457: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  a1, a2 = asarray(a1), asarray(a2)
Out[5]: False
```
So I guess it would be better to test separately `hdus[1].data['test'][0)` and `hdus[1].data['test'][1]`, or make a `np.array(..., dtype=object)`.

cmarmo commented:
> I should have seen that so no blame on you, but it seems we now have a few issues to deal with.

Well a bit of blame on me ... quand-même!
I am really sorry. I will offer a PR ASAP.

cmarmo commented:
I have fixed the issue related to #13621 : do you prefer two different pull requests or a single one when both are fixed?

saimn commented:
@cmarmo - Thanks for the quick answer :). Et pas de soucis, ça arrive. Separate pull requests would probably be easier for the backports.
